orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
       .def("reserve", [](std::vector<OrtValue>* v, const size_t len) { v->reserve(len); })
       .def("shrink_to_fit", [](std::vector<OrtValue>* v) { v->shrink_to_fit(); })
       .def("__len__", [](const std::vector<OrtValue>& v) { return v.size(); })
-      .def("__iter__", [](const std::vector<OrtValue>& v) {
-        return py::make_iterator(v.cbegin(), v.cend());
-      },
-           py::keep_alive<0, 1>())
+      .def(
+          "__iter__", [](const std::vector<OrtValue>& v) {
+            return py::make_iterator(v.cbegin(), v.cend());
+          },
+          py::keep_alive<0, 1>())
       .def("__getitem__", [](const std::vector<OrtValue>& v, const size_t idx) {
         return v.at(idx);
       })
       .def("dlpack_at", [](std::vector<OrtValue>* v, const size_t idx) {
         return py::reinterpret_steal<py::object>(ToDlpack(v->at(idx)));
-      });
+      })
+      .def("to_dlpack", [](const std::vector<OrtValue>& v, py::object to_tensor) -> py::list {
+
+        if (v.size() == 0)
+          return py::list();
+
+        py::list res;
+
+        PyObject* capsule;
```

Comment:
>PyObject* capsule [](http://example.com/codeflow?start=8&length=17)

Using py::capsule is very convenient and succinct.

Reply:
[https://stackoverflow.com/questions/54876346/pybind11-and-stdvector-how-to-free-data-using-capsules](https://stackoverflow.com/questions/54876346/pybind11-and-stdvector-how-to-free-data-using-capsules)